rust/src/index/vector/diskann/builder.rs

```diff
@@ -0,0 +1,494 @@
+            .ok_or(Error::Index(format!("column {} not found", column)))?,
+    );
+    let matrix: MatrixView = vectors.try_into()?;
+    let pq_code = pq.transform(&matrix, metrix_type).await?;
+    let mut graph: GraphBuilder<PQVertex, DiskANNParams> = row_ids
+        .values()
+        .iter()
+        .enumerate()
+        .map(|(i, &row_id)| PQVertex {
+            row_id,
+            pq: as_primitive_array(pq_code.value(i).as_ref()).clone(),
+        })
+        .collect();
+
+    let distribution = Uniform::new(0, batch.num_rows());
+    // Randomly connect to r neighbors.
+    for i in 0..graph.len() {
+        let mut neighbor_ids: HashSet<u32> = graph.neighbors(i).iter().copied().collect();
+
+        while neighbor_ids.len() < r {
+            let neighbor_id = rng.sample(distribution);
+            if neighbor_id != i {
+                neighbor_ids.insert(neighbor_id as u32);
+            }
+        }
+
+        // Make bidirectional connections.
+        {
+            let n = graph.neighbors_mut(i);
+            n.clear();
+            n.extend(neighbor_ids.iter().copied());
+        }
+        {
+            for neighbor_id in neighbor_ids.iter() {
+                graph.add_edge(*neighbor_id as usize, i);
+            }
+        }
+    }
+
+    Ok((graph, matrix))
+}
+
+/// Distance from the query vector to the vector at the given idx.
+fn distance_to(vectors: &MatrixView, query: &[f32], idx: usize) -> Result<f32> {
+    let vector = vectors
+        .row(idx)
+        .ok_or(Error::Index("Invalid row index".to_string()))?;
+
+    let dists = l2_distance(query, vector, vectors.num_columns())?;
+    Ok(dists.value(0))
+}
+
+/// Distance between two vectors in the matrix.
+fn distance(matrix: &MatrixView, i: usize, j: usize) -> Result<f32> {
+    let vector_i = matrix
+        .row(i)
+        .ok_or(Error::Index("Invalid row index".to_string()))?;
+    let vector_j = matrix
+        .row(j)
+        .ok_or(Error::Index("Invalid row index".to_string()))?;
+
+    let dists = l2_distance(vector_i, vector_j, matrix.num_columns())?;
+    Ok(dists.value(0))
+}
+
+/// Greedy search.
+///
+/// Algorithm 1 in the paper.
+///
+/// Parameters:
+/// - start: The starting vertex.
+/// - query: The query vector.
+/// - k: The number of nearest neighbors to return.
+/// - search_size: Search list size, L in the paper.
+async fn greedy_search(
+    graph: &GraphBuilder<PQVertex, DiskANNParams>,
+    vectors: &MatrixView,
+    start: usize,
+    query: &[f32],
+    k: usize,
+    search_size: usize, // L in the paper.
+) -> Result<(Vec<usize>, HashSet<usize>)> {
+    let mut visited: HashSet<usize> = HashSet::new();
+
+    // L in the paper.
+    // A map from distance to vertex id.
+    let mut candidates: BTreeMap<OrderedFloat<f32>, usize> = BTreeMap::new();
+    let mut heap: BinaryHeap<VertexWithDistance> = BinaryHeap::new();
+    let dist = distance_to(vectors, query, start)?;
+    heap.push(VertexWithDistance {
```

Comment:
the heap is never more than one element?

Reply:
so why use the heap at all?

Reply:
Got lost of a few coommits. Refactored and fixed.